server/src/main/java/io/crate/execution/engine/fetch/ReaderBucket.java

```diff
@@ -62,17 +71,30 @@ IntArrayList sortedDocs() {
         return sortedDocs;
     }
 
-    void fetched(Bucket bucket) {
+    /**
+     * @return the bytes added to the ramAccounting instance
+     **/
+    long fetched(Bucket bucket) {
         assert bucket.size() == docs.size()
             : String.format(Locale.ENGLISH, "requested %d docs but got %d", docs.size(), bucket.size());
         assert sortedDocs != null : "sortedDocs() must have been called before fetched()";
 
         Iterator<Row> rowIterator = bucket.iterator();
+        long bytesAccounted = 0;
         for (var cursor : sortedDocs) {
-            docs.put(cursor.value, rowIterator.next().materialize());
+            Object[] cells = rowIterator.next().materialize();
+            bytesAccounted += accountMemory(cells);
```

Comment:
This is a bit of a pity. The `Bucket` is already in memory and accounted for with the request breaker, but we still have to make a copy because `Bucket` may use shared row objects. We also can't do a `iterator.remove()` call or so to at least free objects as we go because `Bucket` is usually a `StreamBucket` as this point and it is backed by a byte stream and de-serialization happens on the fly when the items are accessed. We cannot remove parts of the bytes stream.


I might follow up on this and explore some options how we could avoid the copy operations here or at least reduce the overhead we generate.